### DISCUSSION TITLE
Question about internal (collected) libraries placement and updating

### DISCUSSION THREAD

**Discussion by ThePromidius** [pyinstaller/pyinstaller#8224] [Help]
So I'm thinking of multiple ways to distribute my application.
I have tried making onefile, onedir and now i'm trying out separated and collected libraries + exe.

### Question is, Is it possible to make the exe find the libraries in the user home directory?

My use case is to have the libraries static and maybe move them with a install.bat to the user home directory under my app folder. That way i can distribute low weigh executables.

The reason i'm trying to make it like this, is because the libraries weigh over 4GB (most of it is torch lib). So decompressing each time the apps  gets executed takes a lot of time.

Thanks for your time and hopefully you will be able to guide and help me!

Any feedback or others approaches are welcome! (be honest if my idea is dumb)

To conclude, the way to update these would be to have the install.bat override them if needed

**Comment by ThePromidius** [comment]
Been thinking about it. I guess i could simply move everything under the home directory and just make a symlink or shortcut. Still wanna hear from you if you have better ideas.

**Reply by bwoodsend** [reply]
Adding the libraries' parent directories to `sys.path` at runtime should be enough to make it findable.

> So decompressing each time the apps gets executed takes a lot of time.

Not if you use onedir mode.

> That way i can distribute low weigh executables.

I don't understand the motivation here. The user still needs to download and install the 4GBs of torch libraries so what's the difference? Is it just smaller updates after the first installation that you're after?

> Any feedback or others approaches are welcome!

Like I say, what are you trying to achieve?

> (be honest if my idea is dumb)

I have never seen package management without a package manager not be a disaster so whilst I won't call it dumb, it certainly doesn't sound sensible. Depending on your target audience, just learning to build Python wheels and telling your users to install Python themselves would likely be easier.

**Comment by ThePromidius** [comment]
> Is it just smaller updates after the first installation that you're after?

Yes, pretty much it. For this i'll just optionally release the single exe if the libraries have not been modified from the previous version. It's way faster to download 300mb than 5GB.

>> Any feedback or others approaches are welcome!

> Like I say, what are you trying to achieve?

Well being honest i'm not sure what i want to archieve. That's why i'm exploring the options for now. 
So smaller updates is one. 
Having the code in one place always and user don't have a folder scattered every time its updated (but i kinda self-answered this myself. just move the lib and executable to the app main dir and make a symlink/shortcut)
Was also thinking in how a simple auto-updater could work replacing the small exe

**Reply by bwoodsend** [reply]
I guess I'll reluctantly recommend [PyUpdater](https://pypi.org/project/PyUpdater/) then. I say *reluctantly* because I personally think that any application that writes to its own location or updates itself, thus tripping over its own feet with permission issues caused by editing files that are in use and preventing the user from securing their environment by making the application read only, is an abomination.
